docs/docs/reference/extend/extension-methods.md

```diff
@@ -0,0 +1,260 @@
+---
+layout: doc-page
+title: "Extension Methods"
+---
+
+Extension methods allow one to add methods to a type after the type is defined. Example:
+
+```scala
+case class Circle(x: Double, y: Double, radius: Double)
+
+implicit object CircleOps {
+  def (c: Circle).circumference: Double = c.radius * math.Pi * 2
+}
+```
+
+`CircleOps` adds an extension method `circumference` to values of class `Circle`. Like regular methods, extension methods can be invoked with infix `.`:
+
+```scala
+  val circle = Circle(0, 0, 1)
+  circle.circumference
+```
+
+### Translation of Extension Methods
+
+Extension methods are methods that have a parameter clause in front of the defined
+identifier. They translate to methods where the leading parameter section is moved
+to after the defined identifier. So, the definition of `circumference` above translates
+to the plain method, and can also be invoked as such:
+```scala
+def circumference(c: Circle): Double = c.radius * math.Pi * 2
+
+assert(circle.circumference == CircleOps.circumference(circle))
+```
+
+### Translation of Calls to Extension Methods
+
+The rules for resolving a selection `e.m` are augmented as follows: If `m` is not a
+member of the type `T` of `e`, and there is an implicit value `i`
+in either the current scope or in the implicit scope of `T`, and `i` defines an extension
+method named `m`, then `e.m` is expanded to `i.m(e)`. This expansion is attempted at the time where the compiler also tries an implicit conversion from `T` to a type containing `m`. If there is more than one way
```

Comment:
Is this purely driven by the name `m`, or should we also check `m`'s signature (against any known argument types / expected result type?)

Reply:
> Is this purely driven by the name m, or should we also check m's signature (against any known argument types/shape/arity and the expected result type?)

We should check the whole signature before settling on an extension method.